react/features/mobile/external-api/middleware.js

```diff
@@ -0,0 +1,75 @@
+/* @flow */
+
+import { NativeModules } from 'react-native';
+
+import { getInviteURL } from '../../base/connection';
+import {
+    CONFERENCE_FAILED,
+    CONFERENCE_JOINED,
+    CONFERENCE_LEFT,
+    CONFERENCE_WILL_JOIN,
+    CONFERENCE_WILL_LEAVE
+} from '../../base/conference';
+import { MiddlewareRegistry } from '../../base/redux';
+
+
+/**
+ * Middleware that captures Redux actions and uses the ExternalAPI module to
+ * turn them into native events so the application knows about them.
+ *
+ * @param {Store} store - Redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(store => next => action => {
+    const eventData = {};
+
+    switch (action.type) {
+    case CONFERENCE_FAILED: {
+        eventData.error = action.error;
+        eventData.url = getInviteURL(store.getState());
```

Comment:
1. `inviteURL` in pretty much meant for display and/or public purposes and, consequently, it doesn't contain URL `hash` and `search` params. Which to me makes it kind of less useful to Jitsi Meet SDK consumers because they specify a URL to `JitsiMeetView.loadURL(URL)` but then potentially get another URL in `JitsiMeetView`'s delegate/listener. `locationURL` is closer to what I'd go with.

2. Technically, there's a possibility that Jitsi Meet will rewrite the URL specified to `loadURL`. It may replace the scheme (if the specified URL uses the app-specific scheme), infer other URL properties from the app's `defaultURL`, and plainly replace the `host` and `pathname` for certain hardcoded `host`s.

3. Due to specifics of the current implementation of tapping on the hangup `ToolbarButton` (which are you painfully aware of), `CONFERENCE_WILL_LEAVE` and `CONFERENCE_LEFT` will report an `eventData.url` which is the app's `defaultURL`.

(4. I hope the type of the value of `action.error` is really a `string`. I know the `action`'s `type` is documented like that but it's not really enforced. If it's not a `string`, I'm not sure how useful that would be to the Jitsi Meet SDK consumers.)

Hopefully, the fact that `eventData` is a dictionary/hashmap for Jitsi Meet SDK consumers discourages them from examining it at all.